File: bot/src/gpt/tool/function/ListStickerPacksFunction.ts
import AssistantFunction, {AssistantObject, InvocationContext} from "./AssistantFunction";
import {VkMessage} from "../../../service/VkMessagesService";
import {Context} from "../../../Context";
import {Logger} from "../../../service/LoggingService";

export default class ListStickerPacksFunction implements AssistantFunction {
    private logger!: Logger;

    constructor(private context: Context) {
        context.onReady(() => {
            this.logger = context.loggingService.getRootLogger().newSublogger("ListStickerPacksFunction");
        })
    }

    async call(args: any, message: VkMessage, invocationContext: InvocationContext): Promise<string> {
        const stickerPacks = await this.context.vkStickerPacksOrmService.getEnabled();
        const result = stickerPacks.map(pack => {
            return {
                packId: pack.productId,
                name: pack.name
            };
        });
        this.logger.info(`Listing ${result.length} sticker packs`);
        return "Available sticker packs:\n" + JSON.stringify(result) + "\n";
    }

    getDescription(): string {
        return "Shows a list of available sticker packs. " +
            "Use 'see_sticker_pack' to see all stickers in a pack and decide which one to send. " +
            "The sticker list is mainly for internal use and should not be shown to the user. ";
    }

    getName(): string {
        return "list_sticker_packs";
    }

    getParameters(): AssistantObject | null {
        return null;
    }
}

File: backend/src/main/java/space/davids_digital/sweetie/gpt/tool/function/ListStickerPacksFunction.kt
package space.davids_digital.sweetie.gpt.tool.function

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import space.davids_digital.sweetie.gpt.InvocationContext
import space.davids_digital.sweetie.model.VkMessageModel
import space.davids_digital.sweetie.orm.service.VkStickerPackOrmService

@Component
class ListStickerPacksFunction(
    private val vkStickerPackOrmService: VkStickerPackOrmService
): AssistantFunction<Unit> {
    companion object {
        private val log = LoggerFactory.getLogger(ListStickerPacksFunction::class.java)
    }

    override fun getName() = "list_sticker_packs"
    override fun getDescription() = """
        Shows a list of available sticker packs. 
        "Use 'see_sticker_pack' to see all stickers in a pack and decide which one to send. 
        The sticker list is mainly for internal use and should not be shown to the user. 
    """.trimIndent()
    override fun getParametersClass() = Unit::class

    override suspend fun call(parameters: Unit, message: VkMessageModel, invocationContext: InvocationContext): String {
        val packs = vkStickerPackOrmService.getEnabled()
        log.info("Listing ${packs.size} sticker packs")
        return "Available sticker packs (id, name):\n" + packs.joinToString("\n") { it.id.toString() + ", " + it.name }
    }
}